reference/progressr.options.md
# Options and environment variables used by the 'progressr' packages

Below are environment variables and R options that are used by the
progressr package. Below are all R options that are currently used by
the progressr package.\
\
*WARNING: Note that the names and the default values of these options
may change in future versions of the package. Please use with care until
further notice.*

## Options for controlling progression reporting

- progressr.handlers::

  (function or list of functions) Zero or more progression handlers that
  will report on any progression updates. If empty list, progress
  updates are ignored. If NULL, the default (`handler_txtprogressbar`)
  progression handlers is used. The recommended way to set this option
  is via
  [`handlers()`](https://progressr.futureverse.org/reference/handlers.md).
  (Default: NULL)

## Options for controlling progression handlers

- progressr.clear::

  (logical) If TRUE, any output, typically visual, produced by a
  reporter will be cleared/removed upon completion, if possible.
  (Default: TRUE)

- progressr.enable::

  (logical) If FALSE, then progress is not reported. (Default: TRUE in
  interactive mode, otherwise FALSE)

- progressr.enable_after::

  (numeric) Delay (in seconds) before progression updates are reported.
  (Default: `0.0`)

- progressr.times::

  (numeric) The maximum number of times a handler should report
  progression updates. If zero, then progress is not reported. (Default:
  `+Inf`)

- progressr.interval::

  (numeric) The minimum time (in seconds) between successive progression
  updates from this handler. (Default: `0.0`)

- progressr.intrusiveness::

  (numeric) A non-negative scalar on how intrusive (disruptive) the
  reporter to the user. This multiplicative scalar applies to the
  *interval* and *times* parameters. (Default: `1.0`)\

  progressr.intrusiveness.audio:

  :   (numeric) intrusiveness for auditory progress handlers (Default:
      `5.0`)

  progressr.intrusiveness.file:

  :   (numeric) intrusiveness for file-based progress handlers (Default:
      `5.0`)

  progressr.intrusiveness.gui:

  :   (numeric) intrusiveness for graphical-user-interface progress
      handlers (Default: `1.0`)

  progressr.intrusiveness.notifier:

  :   (numeric) intrusiveness for progress handlers that creates
      notifications (Default: `10.0`)

  progressr.intrusiveness.terminal:

  :   (numeric) intrusiveness for progress handlers that outputs to the
      terminal (Default: `1.0`)

  progressr.intrusiveness.debug:

  :   (numeric) intrusiveness for "debug" progress handlers (Default:
      `0.0`)

## Options for controlling how standard output and conditions are relayed

- progressr.delay_conditions::

  (character vector) condition classes to be captured and relayed at the
  end after any captured standard output is relayed. (Default:
  `c("condition")`)

- progressr.delay_stdout::

  (logical) If TRUE, standard output is captured and relayed at the end
  just before any captured conditions are relayed. (Default: TRUE)

## Options for controlling interrupts

- progressr.interrupts::

  (logical) Controls whether interrupts should be detected or not. If
  FALSE, then interrupts are not detected and progress information is
  generated. (Default: `TRUE`)

- progressr.delay_stdout::

  (logical) If TRUE, standard output is captured and relayed at the end
  just before any captured conditions are relayed. (Default: TRUE)

## Options for debugging progression updates

- progressr.debug::

  (logical) If TRUE, extensive debug messages are generated. (Default:
  FALSE)

## Options for progressr examples and demos

- progressr.demo.delay::

  (numeric) Delay (in seconds) between each iteration of
  [`slow_sum_p()`](https://progressr.futureverse.org/reference/slow_sum_p.md)
  and
  [`slow_sqrt_p()`](https://progressr.futureverse.org/reference/slow_sum_p.md).
  (Default: `1.0` and `0.5`)

- progressr.demo.stdout::

  (logical) If TRUE, then a text is outputted to the standard output per
  element. (Default: `FALSE`)

- progressr.demo.message::

  (logical) If TRUE, then a message is outputted per element. (Default:
  `TRUE`)

- progressr.demo.sticky::

  (logical) If TRUE, then a "sticky" message is outputted every ten
  element. (Default: `TRUE`)

- progressr.demo.progress::

  (logical) If TRUE, then a progressor is created, otherwise a void
  progressor that does nothing is used. (Default: `TRUE`)

## Environment variables that set R options

Some of the above R progressr.\* options can be set by corresponding
environment variable `R_PROGRESSR_*` *when the progressr package is
loaded*. For example, if `R_PROGRESSR_ENABLE = "true"`, then option
progressr.enable is set to `TRUE` (logical). For example, if
`R_PROGRESSR_ENABLE_AFTER = "2.0"`, then option progressr.enable_after
is set to `2.0` (numeric).

## See also

To set R options when R starts (even before the progressr package is
loaded), see the [Startup](https://rdrr.io/r/base/Startup.html) help
page. The [startup](https://cran.r-project.org/package=startup) package
provides a friendly mechanism for configuring R at startup.


NEWS.md
# Version (development version)

 * ...


# Version 1.0.0 [2026-07-03]

## Significant Changes

 * Changed the package license to permissive Apache License (>= 2).

## New Features

 * Add `handler_tutorial()` useful for benchmarking and illustrating
   sequential and parallel processing, e.g. `handlers("tutorial")` and
   `handlers("tutorial", "newline")`.

 * `handler_cli()` gained argument `type` to choose from a predefined
   set of `cli::cli_progress_bar()` `format` strings.

 * `handler_progress()` gained argument `type` to choose from a
   predefined set of `progress::progress_bar()` `format` strings.

 * Renamed `slow_sum()` to `slow_sum_p()` and added `slow_sqrt_p()`.
   Added new basic `slow_sum()` and `slow_sqrt()`.


# Version 0.19.0 [2026-03-18]

## New Features

 * `handlers()` gained support for specifying both progression
   handlers and enabling the global progression handler in one
   call, e.g. `handlers("beepr", "cli", global = TRUE)`.

## Performance

 * `handler_cli()` no longer forces a display update after every
   tick, instead relying on **cli** itself to decide how often to
   refresh the display. This results in much better performance.

## Bug Fixes

 * Buffering and flushing of standard output (stdout) would cause a
   progress bar in the terminal to overwrite flush output that did not
   include a newline. Although it was technically flushed, it would be
   invisible in the terminal. Now buffered stdout is only flushed up
   until the last newline, keeping the remaining output buffered until
   it is safe to flush.

## Deprecated and Defunct

 * Remove `progress()`, which has been defunct since **progressr**
   0.10.0 (2021-12-18).


# Version 0.18.0 [2025-11-05]

## New Features

 * Add `length()` for `progressor` objects, which enables using
   `seq_along()` and `seq()` on progressors.

## Beta Features

 * Add experimental support for `progressor(..., finalize = TRUE)`,
   which will cause the progressor to signal a progression 'shutdown'
   condition when the progressor is garbage collected. This can serve
   as a backstop in yet-to-be-observed cases where a progressor did
   not shut down properly leaving for instance active sinks behind.
   For now, `finalize = FALSE` is the default.

## Bug Fixes

 * It was not possible to create more than one progressor in a
   `with_progress()` call, resulting in additional progressors being
   ignored. Also, warnings on "with_progress() received a progression
   'initiate' request" were produced with **progressr** 0.17.0
   (2025-10-15).
   

# Version 0.17.0 [2025-10-15]

## New Features

 * Now also `with_progress()` warns about receiving progress updates
   after the progressor has finished. Previously, only the global
   'progression' handler would do this.
 
## Bug Fixes

 * Using `with_progress()` together with `handlers(global = TRUE)`
   would produce a warning on "Received a progression 'update' request
   (...) but is not listening to this progressor".

 * Shiny apps running as background jobs in RStudio could fail with an
   error "object 'RStudio.Version' of mode 'function' was not found".
   This error occurred sporadically.
   

# Version 0.16.0 [2025-05-18]

## New Features

 * Add `handler_slowdown()` for slowing down a full progress run a
   certain amount of time (in seconds).

## Workarounds

 * RStudio v2025.05.0 introduced new default handlers for messages and
   warnings for the RStudio Console, which came with a bug causing
   console output to stop working in some cases. This bug was fixed in
   RStudio v2025.05.1. If you run v2025.05.0, **progressr** implements
   a workaround.

## Documentation

 * Replace the long, single introductory vignette with eight separate
   vignettes.

## Bug Fixes

 * Package failed to load in the Positron environment due to an
   "object 'RStudio.Version' of mode 'function' was not found" error.

 * In Positron, messages produced while progress is reported were
   prepended by a newline to standard output, even when there is no
   stdout buffered. This is due to a bug in Positron, which this version
   of the package works around.


# Version 0.15.1 [2024-11-20]

## Bug Fixes

 * Renamed `handler_rpushbullet()` argument `device` to `devices`.

 * Fixed thinko bug in `example("handler_ntfy")`.


# Version 0.15.0 [2024-10-28]

## New Features

 * Add `handler_ntfy()` for reporting on progress via the Ntfy.sh
   Messaging Service using the **ntfy** package.


## Bug Fixes

 * `withProgressShiny()` could produce an `if (config$max_steps ==
   0) : ... argument is of length zero` error.

 * `handlers(new_handlers)` would return `NULL`, instead of `list()`,
   if there were no prior handlers set.

 * `handler_cli(..., format_done = "...", clear = TRUE)` would not
   render the `format_done` message, if set.


# Version 0.14.0 [2023-08-10]

## New Features

 * Progress reporting may be terminated by an interrupt (e.g. user
   presses Ctrl-C, or the process is interrupted externally), or a
   run-time error.  When this happens, the most recent progress update
   is preserved (e.g. a progress bar in the terminal remains), and an
   informative message is displayed (if the progress handler supported
   it). In previous versions, the preservation of the progress and the
   output of the message happened only for interrupts. In this version,
   this happens also for errors.

## Miscellaneous

 * The `progressr.options` help page is now listed in the help index.

## Bug Fixes

 * The 'cli', 'pbcol', 'pbmclapply', 'progress', and 'txtprogressbar'
   handlers did not redraw the progress bar if there was an interrupt,
   which made it a hit or miss whether it was displayed after the
   interruption.
 

# Version 0.13.0 [2023-01-09]

## Significant Changes

 * Now the 'shiny' and 'filesize' handlers are enabled by
   default. Previously, they were only enabled in interactive mode,
   but as these are frequently used also in non-interactive mode, it's

R/options.R
#' Options and environment variables used by the 'progressr' packages
#'
#' Below are environment variables and \R options that are used by the
#' \pkg{progressr} package.
#
#' Below are all \R options that are currently used by the \pkg{progressr} package.\cr
#' \cr
#' \emph{WARNING: Note that the names and the default values of these options may change in future versions of the package.  Please use with care until further notice.}
#'
#'
#' @section Options for controlling progression reporting:
#'
#' \describe{
#'   \item{\option{progressr.handlers}:}{
#'     (function or list of functions)
#'     Zero or more progression handlers that will report on any progression updates.  If empty list, progress updates are ignored.  If NULL, the default (`handler_txtprogressbar`) progression handlers is used.  The recommended way to set this option is via [progressr::handlers()]. (Default: NULL)
#'   }
#' }
#'
#'
#' @section Options for controlling progression handlers:
#'
#' \describe{
#'   \item{\option{progressr.clear}:}{
#'     (logical)
#'     If TRUE, any output, typically visual, produced by a reporter will be cleared/removed upon completion, if possible. (Default: TRUE)
#'   }
#'
#'   \item{\option{progressr.enable}:}{
#'     (logical)
#'     If FALSE, then progress is not reported.
#'     (Default: TRUE in interactive mode, otherwise FALSE)
#'   }
#'
#'   \item{\option{progressr.enable_after}:}{
#'     (numeric)
#'     Delay (in seconds) before progression updates are reported.
#'     (Default: `0.0`)
#'   }
#'
#'   \item{\option{progressr.times}:}{
#'     (numeric)
#'     The maximum number of times a handler should report progression updates. If zero, then progress is not reported.
#'     (Default: `+Inf`)
#'   }
#'
#'   \item{\option{progressr.interval}:}{
#'     (numeric)
#'     The minimum time (in seconds) between successive progression updates from this handler.
#'     (Default: `0.0`)
#'   }
#'
#'   \item{\option{progressr.intrusiveness}:}{
#'     (numeric)
#'     A non-negative scalar on how intrusive (disruptive) the reporter to the user. This multiplicative scalar applies to the _interval_ and _times_ parameters. (Default: `1.0`)\cr
#'   
#'    \describe{
#'      \item{\option{progressr.intrusiveness.audio}:}{(numeric) intrusiveness for auditory progress handlers (Default: `5.0`)}
#'      \item{\option{progressr.intrusiveness.file}:}{(numeric) intrusiveness for file-based progress handlers (Default: `5.0`)}
#'      \item{\option{progressr.intrusiveness.gui}:}{(numeric) intrusiveness for graphical-user-interface progress handlers (Default: `1.0`)}
#'      \item{\option{progressr.intrusiveness.notifier}:}{(numeric) intrusiveness for progress handlers that creates notifications (Default: `10.0`)}
#'      \item{\option{progressr.intrusiveness.terminal}:}{(numeric) intrusiveness for progress handlers that outputs to the terminal (Default: `1.0`)}
#'      \item{\option{progressr.intrusiveness.debug}:}{(numeric) intrusiveness for "debug" progress handlers (Default: `0.0`)}
#'    }
#'   }
#' }
#'
#' @section Options for controlling how standard output and conditions are relayed:
#'
#' \describe{
#'   \item{\option{progressr.delay_conditions}:}{
#'     (character vector)
#'     condition classes to be captured and relayed at the end after any captured standard output is relayed. (Default: `c("condition")`)
#'   }
#'
#'   \item{\option{progressr.delay_stdout}:}{
#'     (logical)
#'     If TRUE, standard output is captured and relayed at the end just before any captured conditions are relayed. (Default: TRUE)
#'   }
#' }
#'
#' @section Options for controlling interrupts:
#'
#' \describe{
#'   \item{\option{progressr.interrupts}:}{
#'     (logical)
#'     Controls whether interrupts should be detected or not.
#'     If FALSE, then interrupts are not detected and progress information
#'     is generated. (Default: `TRUE`)
#'   }
#'
#'   \item{\option{progressr.delay_stdout}:}{
#'     (logical)
#'     If TRUE, standard output is captured and relayed at the end just before any captured conditions are relayed. (Default: TRUE)
#'   }
#' }
#'
#'
#' @section Options for debugging progression updates:
#'
#' \describe{
#'  \item{\option{progressr.debug}:}{(logical) If TRUE, extensive debug messages are generated. (Default: FALSE)}
#' }
#'
#'
#' @section Options for progressr examples and demos:
#'
#' \describe{
#'  \item{\option{progressr.demo.delay}:}{(numeric) Delay (in seconds) between each iteration of [slow_sum_p()] and [slow_sqrt_p()]. (Default: `1.0` and `0.5`)}
#'
#'  \item{\option{progressr.demo.stdout}:}{(logical) If TRUE, then a text is outputted to the standard output per element. (Default: `FALSE`)}
#'
#'  \item{\option{progressr.demo.message}:}{(logical) If TRUE, then a message is outputted per element. (Default: `TRUE`)}
#'
#'  \item{\option{progressr.demo.sticky}:}{(logical) If TRUE, then a "sticky" message is outputted every ten element. (Default: `TRUE`)}
#'
#'  \item{\option{progressr.demo.progress}:}{(logical) If TRUE, then a progressor is created, otherwise a void progressor that does nothing is used. (Default: `TRUE`)}
#' }
#'
#' @section Environment variables that set R options:
#' Some of the above \R \option{progressr.*} options can be set by corresponding
#' environment variable \env{R_PROGRESSR_*} _when the \pkg{progressr} package
#' is loaded_.
#' For example, if `R_PROGRESSR_ENABLE = "true"`, then option
#' \option{progressr.enable} is set to `TRUE` (logical).
#' For example, if `R_PROGRESSR_ENABLE_AFTER = "2.0"`, then option
#' \option{progressr.enable_after} is set to `2.0` (numeric).
#'
#' @seealso
#' To set \R options when \R starts (even before the \pkg{progressr} package is loaded), see the \link[base]{Startup} help page.  The \href{https://cran.r-project.org/package=startup}{\pkg{startup}} package provides a friendly mechanism for configuring \R at startup.
#'
#' @aliases progressr.clear
#' @aliases progressr.debug
#' @aliases progressr.demo.delay
#' @aliases progressr.demo.stdout progressr.demo.message progressr.demo.sticky progressr.demo.progress
#' @aliases progressr.delay_stdout progressr.delay_conditions
#' @aliases progressr.enable progressr.enable_after
#' @aliases progressr.interrupts
#' @aliases progressr.interval
#' @aliases progressr.intrusiveness
#' @aliases progressr.intrusiveness.audio
#' @aliases progressr.intrusiveness.debug
#' @aliases progressr.intrusiveness.file
#' @aliases progressr.intrusiveness.gui
#' @aliases progressr.intrusiveness.notifier
#' @aliases progressr.intrusiveness.terminal
#' @aliases progressr.handlers
#' @aliases progressr.times
#'
#' @name progressr.options
NULL


get_package_option <- function(name, default = NULL, package = .packageName) {
  if (!is.null(package)) {
    name <- paste(package, name, sep = ".")
  }
  getOption(name, default = default)
}

# Set an R option from an environment variable
update_package_option <- function(name, mode = "character", default = NULL, package = .packageName, split = NULL, trim = TRUE, disallow = c("NA"), force = FALSE, debug = FALSE) {
  if (!is.null(package)) {
    name <- paste(package, name, sep = ".")
  }

  mdebugf("Set package option %s", sQuote(name))

  ## Already set? Nothing to do?
  value <- getOption(name, NULL)
  if (!force && !is.null(value)) {
    mdebugf("Already set: %s", sQuote(value))
    return(getOption(name))
  }

  ## name="Pkg.foo.Bar" => env="R_PKG_FOO_BAR"
  env <- gsub(".", "_", toupper(name), fixed = TRUE)
  env <- paste("R_", env, sep = "")

  env_value <- value <- Sys.getenv(env, unset = NA_character_)
  if (is.na(value)) {  
    if (debug) mdebugf("Environment variable %s not set", sQuote(env))
    
    ## Nothing more to do?
    if (is.null(default)) return(getOption(name))

    if (debug) mdebugf("Use argument 'default': ", sQuote(default))
    value <- default
  }

  if (debug) mdebugf("%s=%s", env, sQuote(value))

  ## Trim?
  if (trim) value <- trim(value)

  ## Nothing to do?
  if (!nzchar(value)) return(getOption(name, default = default))

  ## Split?
  if (!is.null(split)) {
    value <- strsplit(value, split = split, fixed = TRUE)
    value <- unlist(value, use.names = FALSE)
    if (trim) value <- trim(value)
  }

  ## Coerce?
  mode0 <- storage.mode(value)
  if (mode0 != mode) {
    suppressWarnings({
      storage.mode(value) <- mode
    })
    if (debug) {
      mdebugf("Coercing from %s to %s: %s", mode0, mode, commaq(value))
    }
  }

  if (length(disallow) > 0) {
    if ("NA" %in% disallow) {
      if (any(is.na(value))) {
        stop(sprintf("Coercing environment variable %s=%s to %s would result in missing values for option %s: %s", sQuote(env), sQuote(env_value), sQuote(mode), sQuote(name), commaq(value)))
      }
    }
    if (is.numeric(value)) {
      if ("non-positive" %in% disallow) {
        if (any(value <= 0, na.rm = TRUE)) {
          stop(sprintf("Environment variable %s=%s specifies a non-positive value for option %s: %s", sQuote(env), sQuote(env_value), sQuote(name), commaq(value)))
        }
      }
      if ("negative" %in% disallow) {
        if (any(value < 0, na.rm = TRUE)) {
          stop(sprintf("Environment variable %s=%s specifies a negative value for option %s: %s", sQuote(env), sQuote(env_value), sQuote(name), commaq(value)))
        }
      }
    }
  }
  
  if (debug) {
    mdebugf("=> options(%s = %s) [n=%d, mode=%s]",
            dQuote(name), commaq(value),
            length(value), storage.mode(value))
  }

  do.call(options, args = structure(list(value), names = name))
  
  getOption(name, default = default)
}


## Set package options based on environment variables
update_package_options <- function(debug = FALSE) {
  update_package_option("demo.delay", mode = "numeric", debug = debug)
  update_package_option("demo.stdout", mode = "logical", debug = debug)
  update_package_option("demo.message", mode = "logical", debug = debug)
  update_package_option("demo.sticky", mode = "logical", debug = debug)
  update_package_option("demo.progress", mode = "logical", debug = debug)

  ## make_progression_handler() arguments
  update_package_option("clear", mode = "logical", default = TRUE, debug = debug)

  ## Special case: Support R_PROGRESSR_ENABLE=interactive
  value <- Sys.getenv("R_PROGRESSR_ENABLE", "")
  if (value == "interactive") {
    options(progressr.enable = interactive())
  } else {
    update_package_option("enable", mode = "logical", default = interactive(), debug = debug)
  }

  update_package_option("enable_after", mode = "numeric", default = 0.0, debug = debug)
  update_package_option("interval", mode = "numeric", default = 0.0, debug = debug)
  update_package_option("times", mode = "numeric", default = +Inf, debug = debug)
  update_package_option("interrupts", mode = "logical", default = TRUE, debug = debug)

  ## Life-cycle, e.g. deprecation an defunct
  update_package_option("lifecycle.progress", mode = "character", default = "deprecated", debug = debug)

  ## However, not used
  update_package_option("global.handler", mode = "logical", debug = debug)

  ## For RStudio users
  update_package_option("rstudio.patch", mode = "logical", debug = debug)
}
